cpp/src/arrow/compute/kernels/scalar_temporal_test.cc

```diff
@@ -976,6 +976,23 @@ TEST_F(ScalarTemporalTest, TestTemporalDifference) {
   }
 }
 
+TEST_F(ScalarTemporalTest, TestTemporalSubtractDuration) {
+  std::string op = "subtract";
+
+  for (auto u : TimeUnit::values()) {
+    auto unit = duration(u);
+    CheckScalarBinary(op,
+                      ArrayFromJSON(unit, times_s2),
+                      ArrayFromJSON(unit, times_s),
+                      ArrayFromJSON(unit, seconds_between_time));
+  }
+
+  EXPECT_RAISES_WITH_MESSAGE_THAT(
+      NotImplemented, testing::HasSubstr("no kernel matching input types"),
+      Subtract(ArrayFromJSON(duration(TimeUnit::SECOND), times_s),
+               ArrayFromJSON(duration(TimeUnit::MILLI), times_s)));
+}
+
```

Comment:
Add tests with valid values.

Reply:
Do you mean [like these](https://github.com/apache/arrow/blob/b3cff8f82fd43d3cbbed841f40bb8a3ca0e274e1/cpp/src/arrow/compute/kernels/scalar_temporal_test.cc#L982-L988)?

I'd like to enable cases like `subtract(duration[ms], duration[us]) -> duration[us]`. I'll ping when it's done.

Reply:
Ok, this now covers cases like `subtract(duration[ms], duration[us]) -> duration[us]`.